tasmota/xdrv_01_webserver.ino

```diff
@@ -3098,14 +3098,86 @@ int WebSend(char *buffer)
       }
       url += F("cmnd=");                      // url = |http://192.168.178.86/cm?cmnd=| or |http://192.168.178.86/cm?user=admin&password=joker&cmnd=|
     }
-    url += command;                           // url = |http://192.168.178.86/cm?cmnd=POWER1 ON|
+    url += command;
+    url += F(" GET");                         // url = |http://192.168.178.86/cm?cmnd=POWER1 ON|
 
     DEBUG_CORE_LOG(PSTR("WEB: Uri |%s|"), url.c_str());
+    status = WebQuery(const_cast<char*>(url.c_str()));
+  }
+  return status;
+}
+
+int WebQuery(char *buffer)
+{
+  // http://192.168.1.1/path GET                                         -> Sends HTTP GET http://192.168.1.1/path
+  // http://192.168.1.1/path POST {"some":"message"}                     -> Sends HTTP POST to http://192.168.1.1/path with body {"some":"message"}
+  // http://192.168.1.1/path PUT [Autorization: Bearer abcdxyz] potato   -> Sends HTTP PUT to http://192.168.1.1/path with authorization header and body "potato"
+  // http://192.168.1.1/path PATCH patchInfo                             -> Sends HTTP PATCH to http://192.168.1.1/path with body "potato"
+
+  // Valid HTTP Commands: GET, POST, PUT, and PATCH
+  // An unlimited number of headers can be sent per request, and a body can be sent for all command types
+  // The body will be ignored if sending a GET command
+
+  int status = WEBCMND_WRONG_PARAMETERS;
+  int http_code;
+  WiFiClient http_client;
+  HTTPClient http;
+  String methodStr;
+  String bodyStr;
+  char *url;
+  char *header;
+  char *headerBody;
+  char *body;
+  char *method;
+  char *temp;
+  bool headerFound;
+
+  url = strtok_r(buffer, " ", &temp);
+  method = strtok_r(temp, " ", &temp);
+
+  if(url && method) {
+    if (http.begin(http_client, UrlEncode(url))) {
+      if(temp) { //There is a body and/or header
+        if(temp[0] == '[') { //Header information was sent; decode it
+          temp += 1;
+          temp = strtok_r(temp, "]", &body);
+          headerFound = true;
+          while(headerFound) {
+            header = strtok_r(temp, ":", &temp);
+            if(header) {
+              headerBody = strtok_r(temp, "|", &temp);
+              if(headerBody) {
+                http.addHeader(header, headerBody);
+              }
+              else headerFound = false;
+            }
+            else headerFound = false;
+          }
+        }
+        else { //No header information was sent, but there was a body
+          body = temp;
+        }
+      }
+      
+      //No body sent; define empty string for the body
+      else {
+        body = (char*)malloc(1);
```

Comment:
I'm afraid that's a memory leak, I don't think this is ever freed.

Reply:
Thanks for the feedback! I missed that. It looks like Theo fixed this in his implementation